src/Juvix/Compiler/Nockma/Translation/FromAsm.hs

```diff
@@ -0,0 +1,1048 @@
+module Juvix.Compiler.Nockma.Translation.FromAsm where
+
+import Juvix.Compiler.Asm.Data.InfoTable qualified as Asm
+import Juvix.Compiler.Nockma.Evaluator
+import Juvix.Compiler.Nockma.Pretty
+import Juvix.Compiler.Nockma.Stdlib
+import Juvix.Compiler.Pipeline.EntryPoint
+import Juvix.Prelude hiding (Atom, Path)
+
+type UserFunctionId = Symbol
+
+data FunctionId
+  = UserFunction UserFunctionId
+  | BuiltinFunction BuiltinFunctionId
+  deriving stock (Generic, Eq)
+
+instance Hashable FunctionId
+
+data BuiltinFunctionId
+  = BuiltinPow2Go
+  | BuiltinPow2
+  | BuiltinAppendRights
+  deriving stock (Eq, Enum, Bounded, Generic)
+
+instance Hashable BuiltinFunctionId
+
+newtype CompilerOptions = CompilerOptions
+  {_compilerOptionsEnableTrace :: Bool}
+
+fromEntryPoint :: EntryPoint -> CompilerOptions
+fromEntryPoint EntryPoint {..} =
+  CompilerOptions
+    { _compilerOptionsEnableTrace = _entryPointDebug
+    }
+
+data FunctionInfo = FunctionInfo
+  { _functionInfoPath :: Path,
+    _functionInfoArity :: Natural
+  }
+
+data CompilerCtx = CompilerCtx
+  { _compilerFunctionInfos :: HashMap FunctionId FunctionInfo,
+    _compilerConstructorArities :: ConstructorArities,
+    _compilerOptions :: CompilerOptions
+  }
+
+type ConstructorArities = HashMap Asm.Tag Natural
+
+type Offset = Natural
+
+data CompilerFunction = CompilerFunction
+  { _compilerFunctionName :: FunctionId,
+    _compilerFunctionArity :: Natural,
+    _compilerFunction :: Sem '[Compiler, Reader CompilerCtx] ()
+  }
+
+data StackId
+  = CurrentFunction
```

Comment:
What do you need to store the current function for? "Current function" is never referenced in JuvixAsm - functions are always referenced by name.